### PR TITLE
Improv bluetooth sample

### PR DESCRIPTION
**Pull request by AdrianSoundy**
## Description
Bluetooth sample for IMPROV Wifi provisioning 

## Types of changes

- [ ] Improvement (non-breaking change that improves a sample)
- [ ] Bug fix (fixes an issue with a current sample)
- [x] New Sample (adds a new sample)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

Signed-off-by: adriansoundy <adriansoundy@gmail.com>

### CODE REVIEW

File: samples/Bluetooth/ImprovWifi/Improv.cs [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+using System.Net.NetworkInformation;
+using System.Device.Wifi;
+using nanoFramework.Networking;
+
+namespace ImprovWifi
+{
+    public class Improv
+    {
+        /// <summary>
+        /// Improv error states
```
Comment (josesimoes) [inline samples/Bluetooth/ImprovWifi/Improv.cs:19]: Missing . at the end of the comment sentence. (others follow)

File: samples/Bluetooth/ImprovWifi/Improv.cs [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+        /// </summary>
+        public enum ImprovError
+        {
+            /// <summary>
+            /// This shows there is no current error state.
+            /// </summary>
+            NO_ERROR,
```
Comment (josesimoes) [inline samples/Bluetooth/ImprovWifi/Improv.cs:26]: We should be using the camel case convention in enum items too...

File: samples/Bluetooth/ImprovWifi/Improv.cs [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+            if (characteristicResult.Error != BluetoothError.Success)
+            {
+                // An error occurred.
+                return characteristicResult.Error;
+            }
+
+            return BluetoothError.Success;
+        }
+
+
```
Comment (josesimoes) [inline samples/Bluetooth/ImprovWifi/Improv.cs:302]: extra line

File: samples/Bluetooth/ImprovWifi/Improv.cs [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+        #region Characteristic event handlers
+        private Buffer GetByteBuffer(byte value)
```
Comment (josesimoes) [inline samples/Bluetooth/ImprovWifi/Improv.cs:304]: missing line

File: samples/Bluetooth/ImprovWifi/Improv.cs [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+            }
+
+            // Respond if Write requires response
+            if (request.Option == GattWriteOption.WriteWithResponse)
+            {
+                request.Respond();
+            }
+        }
+
+        Buffer SetupRpcResult()
```
Comment (josesimoes) [inline samples/Bluetooth/ImprovWifi/Improv.cs:447]: Missing visibility, presume it should be `private`.

File: samples/Bluetooth/ImprovWifi/Improv.cs [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+
+        private void CheckSum(ref byte cs, byte[] bytes)
+        {
+            foreach (byte b in bytes)
+            {
+                cs += b;
+            }
+        }
+
+
```
Comment (josesimoes) [inline samples/Bluetooth/ImprovWifi/Improv.cs:497]: extra line

File: samples/Bluetooth/ImprovWifi/README.md [hunk 7]
```diff
@@ -0,0 +1,17 @@
+# Improv Wifi provisioning
+
+This sample shows the use of the *Improv class* to provision the Wifi credentials for an ESP32 device via Bluetooth LE. 
+
+The device will advertise with the name "Improv sample" and support the *Improv Service* allowing the Wifi Credentials to be setup directly from the Improv test page.
+or any other web page.
+
+This is an initial working version of the *Improv* class showing what can be done. 
+Probably later this class can be moved to a separate repo. so it has its own nu-get to make it easy to include in you own application.
```
Comment (josesimoes) [inline samples/Bluetooth/ImprovWifi/README.md:9]: Maybe something like this... 

```suggestion
Probably later this class can be moved to a separate repo, so it has its own NuGet package to make it easy to consume by applications.
```

**Review by josesimoes — COMMENTED** [review]
Great sample! Well done. 👏🏻 
When this is published we should make some "noise" about supporting this now.

**Comment by AdrianSoundy** [comment]
I agree about promoting it once its had some further testing for edge cases.

**Comment by josesimoes** [comment]
> I am away for next 2/3 weeks so not sure how much i can do. Just wanted to get everything i have bee working on committed.

@AdrianSoundy that's great! Actually both. Enjoy your time off. 😄

**Review by josesimoes — APPROVED** [review]
Great sample! Thanks.

**Pull request merged**